internal/pkg/cli/file/watch.go

```diff
@@ -0,0 +1,130 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package file
+
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// RecursiveWatcher wraps an fsnotify Watcher to recursively watch all files in a directory.
+type RecursiveWatcher struct {
+	fsnotifyWatcher *fsnotify.Watcher
+	done            chan struct{}
+	closed          bool
+	events          chan fsnotify.Event
+	errors          chan error
+}
+
+// NewRecursiveWatcher returns a RecursiveWatcher which notifies when changes are made to files inside a recursive directory tree.
+func NewRecursiveWatcher() (*RecursiveWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	rw := &RecursiveWatcher{
+		events:          make(chan fsnotify.Event),
+		errors:          make(chan error),
+		fsnotifyWatcher: watcher,
+		done:            make(chan struct{}),
+		closed:          false,
+	}
+
+	go rw.start()
+
+	return rw, nil
+}
+
+// Add recursively adds a directory tree to the list of watched files.
+func (rw *RecursiveWatcher) Add(path string) error {
+	if rw.closed {
+		return fsnotify.ErrClosed
+	}
+	if err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return rw.fsnotifyWatcher.Add(path)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Remove recursively removes a directory tree from the list of watched files.
+func (rw *RecursiveWatcher) Remove(path string) error {
+	if rw.closed {
+		return nil
+	}
+	if err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return rw.fsnotifyWatcher.Remove(path)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Events returns the events channel.
+func (rw *RecursiveWatcher) Events() <-chan fsnotify.Event {
+	return rw.events
+}
+
+// Errors returns the errors channel.
+func (rw *RecursiveWatcher) Errors() <-chan error {
+	return rw.errors
+}
+
+// Close closes the RecursiveWatcher.
+func (rw *RecursiveWatcher) Close() error {
+	rw.closed = true
+	close(rw.done)
+	return rw.fsnotifyWatcher.Close()
+}
+
+func (rw *RecursiveWatcher) start() {
+	for {
+		select {
+		case event := <-rw.fsnotifyWatcher.Events:
+			info, err := os.Stat(event.Name)
+			if err != nil {
+				break
+			}
+			if info.IsDir() {
+				switch event.Op {
+				case fsnotify.Create:
+					err := rw.Add(event.Name)
```

Comment:
Nope, `rw.Add` calls the underlying `Add` function in the `fsnotify.Watcher`. This function can be both absolute or relative. I think this can only become an issue if the relative position being called from changes. This should error out though, and because `--watch` programmatically calls this, it's just reusing the directory provided by `ws.Path()`, so we'll never run into this issue here anyways.